lib/cocoapods/sources_manager.rb

```diff
@@ -41,7 +41,15 @@ def aggregate_for_dependency(dependency)
       #
       def sources(names)
         dirs = names.map { |name| source_dir(name) }
-        dirs.map { |repo| Source.new(repo) }
+        dirs.map do |repo|
+          if Pathname(repo).basename == 'master'
```

Comment:
Given that `Source::Aggregate` also instantiates sources, I would be in favor of moving `MasterSource` into CocoaPods/Core and adding a static method e.g. `Source[path]`, which does this check here. Everywhere where we are currently using `Source.new`, we could use this new method then instead.